plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -314,13 +322,33 @@ void deleteBlobs(Collection<String> blobNames) throws IOException {
             return;
         }
         final List<BlobId> blobIdsToDelete = blobNames.stream().map(blob -> BlobId.of(bucketName, blob)).collect(Collectors.toList());
-        final List<Boolean> deletedStatuses = SocketAccess.doPrivilegedIOException(() -> client().delete(blobIdsToDelete));
-        assert blobIdsToDelete.size() == deletedStatuses.size();
+        final List<Boolean> deletedStatuses = new ArrayList<>();
+        final int deleteCount = blobIdsToDelete.size();
+        SocketAccess.doPrivilegedIOException(() -> {
+            final int batches = deleteCount / MAX_BATCH_SIZE + (deleteCount % MAX_BATCH_SIZE == 0 ? 0 : 1);
+            for (int i = 0; i < batches - 1; ++i) {
+                deletedStatuses.addAll(client().delete(blobIdsToDelete.subList(i * MAX_BATCH_SIZE, (i + 1) * MAX_BATCH_SIZE)));
+            }
+            deletedStatuses.addAll(client().delete(blobIdsToDelete.subList(batches - 1, deleteCount)));
+            return null;
+        });
+        assert deleteCount == deletedStatuses.size();
         boolean failed = false;
         for (int i = 0; i < blobIdsToDelete.size(); i++) {
             if (deletedStatuses.get(i) == false) {
-                logger.error("Failed to delete blob [{}] in bucket [{}]", blobIdsToDelete.get(i).getName(), bucketName);
-                failed = true;
+                final BlobId blobId = blobIdsToDelete.get(i);
+                try {
+                    Blob blob = SocketAccess.doPrivilegedIOException(() -> client().get(blobId));
```

Comment:
I'd rather trust GCP and do not perform any additional checks.

Reply:
GCP doesn't give us a different response for errors and didn't exist, we just get a list of `Boolean` back and I think given current stale data issues we probably want to at least log the deletes that didn't work out?

Reply:
Ok, so google will return "false" if the file does not exist or if there is a failure during removal. However, do we normally try to delete files that initially do not exist? If not, why not just to check if the status is "false" and log an error in this case?

Reply:
> However, do we normally try to delete files that initially do not exist?

There are a few spots where this can happen yes. See reproducer by Yannick here https://github.com/elastic/elasticsearch/pull/40188#discussion_r267290825 for example.

Also, now now that we are fixing the delete for master failover cases we will likely have  a few more of these cases since redundant deletes are cheaper to send than first checking for the file and then deleting (no cost for a `DELETE` call on e.g. S3 and always just one API call per blob).